web/website/content/posts/2023-03-14-pi-day.md

```diff
@@ -0,0 +1,370 @@
+they use a "recursive" (i.e. self-referential) syntax to provide a looping
+construct in SQL.
+
+[PRQL](www.prql-lang.org) is a modern, functional query language for
+transforming data. One of its goals is to simplify working with data wherever
+you can currently use SQL. As such it transpiles to SQL while making available
+modern ergonomics such as
+[f-strings](https://prql-lang.org/book/language-features/f-strings.html), as
+well as not so modern features such as
+[functions](https://prql-lang.org/book/queries/functions.html).
+
+Given that the underlying semantics of Recursive CTEs are really about iteration
+or "looping", we have called this feature `loop` in PRQL.
+
+## Introducing `loop`
+
+The way Recursive CTEs work in SQL (provide Teggy paper reference), is that they
+consist of two parts, an `initial_query` and an `update_query`. First the
+`initial_query` is executed and then the rows produced are fed to `update_query`
+which is applied to the result set. The `update_query` is then iteratively
+applied to the rows produced in the last iteration until no more rows are
+produced, at which point iteration stops. (For a great review of this as well as
+some interesting proposals to extend the semantics of Recursive CTEs in SQL, see
+the paper
+["A Fix for the Fixation on Fixpoints" by Denis Hirn and Torsten Grust](https://www.cidrdb.org/cidr2023/papers/p14-hirn.pdf).)
+
+This behavior can be expressed with following pseudo-code:
+
+```python
+def loop(step, initial_query):
+    result = []
+    current = initial_query()
+    while current is not empty:
+        result = append(result, current)
+        current = update_query(current)
+
+    return result
+```
+
+The minimal `loop` example from the documentation in the
+[PRQL book](https://prql-lang.org/book/language-features/standard-library/loop.html)
+is:
+
+```elm=
+from_text format:json '[{"n": 1 }]'
+loop (
+    select n = n+1
+    filter n<=3
+)
+```
+
+Here we use a PRQL utility function `from_text` to conveniently turn a JSON
+representation of some example data into a SQL table (`from_text` currently also
+accepts CSV input).
+
+Initially a row with `{"n":1}` is fed in. Then the update query is applied to
+this which in this case just increments `n` by one and filters the result set to
+only the rows where `n` is less than or equal to three. This produces `{"n":2}`
+in the next step and `{"n":3}` after that. On the next iteration `{"n":4}` is
+produced but that is eliminated by the filter condition. Since that leaves no
+new rows, iteration stops.
+
+If you try the query above in the
+[PRQL Playground](www.prql-lang.org/playground), the result set you get (in the
+"output.arrow" tab) is:
+
+| n   |
+| --- |
+| 1   |
+| 2   |
+| 3   |
+
+## Fibonacci Numbers
+
+Inspired by this, let's try to calculate Fibonnacci numbers which are often one
+of the first examples when recursion is introduced:
+
+```elm=
+from_text format:json '[{"a":1, "b":1}]'
+loop (
+    select [a=b, b=s"a"+b]
```

Comment:
Is it worth mentioning that the s-string is a workaround?

Reply:
Yes, you're right. I'll add that.